File: code/App/algo.py

from queue import Queue

kevin = "Kevin Bacon"
graph = {}

def read_graph_from_txt(filename):
    try:
        file = open(filename)
    except FileNotFoundError:
        print('File', filename, 'is not found!')
    else:
        with file:
            for line in file:
                actor, *coactors = map(str.strip, line.split(','))
                graph[actor] = coactors


def find_degree(actor):
    if actor == kevin:
        return 0

    q_actor_with_degree = Queue()
    visited_actors = set()

    q_actor_with_degree.put((kevin, 0))

    while not q_actor_with_degree.empty():
        current_actor, current_degree = q_actor_with_degree.get()

        for actor_connected_to_current in graph.get(current_actor, []):
            if actor_connected_to_current == actor:
                return current_degree + 1

            if actor_connected_to_current not in visited_actors:
                visited_actors.add(actor_connected_to_current)
                q_actor_with_degree.put((actor_connected_to_current, current_degree + 1))

    return -1

File: code/App/algo.cc
#include<iostream>
#include<fstream>
#include<sstream>
#include<vector>
#include<unordered_map>
#include<unordered_set>
#include<queue>

const std::string kevin = "Diana Ross";

std::unordered_map<std::string, std::vector<std::string>> graph;

void readGraphFromTxt(std::string filename) {
    std::ifstream file(filename);
    if (!file.is_open()) {
        std::cerr << "err" << filename << std::endl;
        return;
    }

    std::string line;
    while (std::getline(file, line)) {
        std::istringstream iss(line);
        std::string actor, coactor;

        std::getline(iss, actor, ',');
        std::cout << actor << std::endl;
        while (std::getline(iss, coactor, ',')) {
            graph[actor].push_back(coactor);
            std::cout << coactor <<",";
        }
        std::cout << std::endl;
    }
    

    file.close();
};

int findDegree(std::string actor) {
    if (actor == kevin) {
        return 0;
    }

    std::queue<std::pair<std::string, int>> qActorWithDegree;
    std::unordered_set<std::string> visitedActors;

    qActorWithDegree.push({kevin, 0});

    while(!qActorWithDegree.empty()) {
        std::pair<std::string, int> current = qActorWithDegree.front();
        qActorWithDegree.pop();

        std::string currentActor = current.first;
        int currentDegree = current.second;

        for (std::string actorConnectedToCurrent : graph[currentActor]) {
            if (actorConnectedToCurrent == kevin) {
                return currentDegree;
            }

            if (visitedActors.find(actorConnectedToCurrent) == visitedActors.end()) {
                visitedActors.insert(actorConnectedToCurrent);
                qActorWithDegree.push({actorConnectedToCurrent, currentDegree + 1});
            }
        }
    }

    return -1;
};

int main() {
    readGraphFromTxt("o2.txt");
    //std::string userActor;

    //std::cout << "your actor: ";
    //std::cin >> userActor;
    auto userActor = "Michael Jackson";

    std::cout << "degree: " << findDegree(userActor) << std::endl;


    return 0;
};